### PR TITLE
Mahout 1999- Automating Multi-Artifact Build Process

### PR DESCRIPTION
**Pull request by rawkintrevo**
### Purpose of PR:
Automating Multi-Artifact Release Process

check by clearing `.m2` cache then running:
`mvn clean install -Pall-scala,all-spark,viennacl,viennacl-omp`

### Important ToDos
Please mark each with an "x"
- [x] A JIRA ticket exists (if not, please create this first)[https://issues.apache.org/jira/browse/ZEPPELIN/]
- [x] Title of PR is "MAHOUT-XXXX Brief Description of Changes" where XXXX is the JIRA number.
- [x] Created unit tests where appropriate
- [x] Added licenses correct on newly added files
- [x] Assigned JIRA to self
- [x] Added documentation in scala docs/java docs, and to website
- [x] Successfully built and ran all unit tests, verified that all tests pass locally.

If all of these things aren't complete, but you still feel it is
appropriate to open a PR, please add [WIP] after MAHOUT-XXXX before the
descriptions- e.g. "MAHOUT-XXXX [WIP] Description of Change"

Does this change break earlier versions?

Is this the beginning of a larger project for which a feature branch should be made?

### CODE REVIEW

**Comment by rawkintrevo** [comment]
This is old magic.  

Would love to get a review if possible from @bimargulies @srowen @smarthi or anyone else who might have had a hand in creating or modding this section of the code base before me. 

Thanks in advance.

File: distribution/pom.xml [hunk 1]
```diff
@@ -79,6 +110,38 @@
         <mahout.skip.distribution>false</mahout.skip.distribution>
       </properties>
     </profile>
+
+    <profile>
+      <id>scala-2.10</id>
```
Comment (rawkintrevo) [inline distribution/pom.xml]: These are unneeded I've dropped.

**Comment by andrewpalumbo** [comment]
Possibly @sslavic may have some ideas if he is available.



Sent from my Verizon Wireless 4G LTE smartphone


-------- Original message --------

**Comment by sslavic** [comment]
@rawkintrevo making release step includes running Maven release plugin, prepare and perform goals. Perform runs a build (of of the sources prepared by prepare, so with non-snapshot version set, and with all changes committed), creates artifacts, and uploads/deploys artifacts on Apache artifact repository - all in a single run. Maybe I missed something in the PR, but I guess there's no code that attaches artifacts previously built with this new shell script, to the Maven build running "perform" - so those artifacts will not be published on Apache Maven artifact repository. I guess the goal is to have all variants (scala, spark) and all artifacts published.

Because of that, when solving same problem in the past I opted for switching from Maven to sbt (see https://issues.apache.org/jira/browse/MAHOUT-1654 and https://github.com/sslavic/mahout/tree/MAHOUT-1654 )

sbt is aware of Scala binary compatibility "feature" and can natively in single build run release all different variants of a (multi-module) project for different Scala versions, and release all of them. Maven has no good support for this Scala specific - just like you do in this new script, you have to run Maven multiple times with different profiles active. sbt does that too internally, it just keeps the context, that all the different internal builds with different settings are all part of one big build execution, and ends up releasing all the artifacts.

**Comment by rawkintrevo** [comment]
thanks @sslavic ! 

The idea here is yes, Maven is run multiple times, then the assembly plugin picks each combination up and rolls it into its own tar ball (which I presumed would get picked up with the release-plugin, but the only way to test that theory as far as I could tell is try it once it was merged, as it won't deploy if the local code doesn't match the code in Master). 

With this sbt you have, would it deploy one big tarball with all of the scala versions or multiple tarballs one for each version?

**Comment by sslavic** [comment]
New script would have to be run between release:prepare and release:perform, so that artifacts script creates and packages in tarball are referencing release and not snapshot versions of submodule artifacts.

Problem is, tarball is not only thing being published and used. Every Mahout submodule (typically it's jar and pom) gets to be published as part of the Maven release. If I'm not mistaken the new script, assembly plugin configuration and Maven build configuration in general as they are now in the PR, do not make and publish all the different pom variants.

Publishing individual submodule POMs and binaries independently from distribution archive is needed by Mahout users which use Mahout as library, so people which declare and pull the Mahout module artifacts from Maven Central / Apache repo as dependencies in their Maven/Gradle/sbt/... builds. If this class of users is not supported, and only distribution tar ball is thing that needs to published and used then new script based release might work.

With sbt's cross Scala binary version build support, every submodules pom and binary would be published in n variants, where n is number of supported Scala binary compatible versions - e.g. 2 variants, for 2.11 and 2.12 Scala which are typically in demand these days, and 2.10 and older support is largely being removed.
Branch with switch from Maven to sbt is outdated and was never 100% complete/tested (e.g. the release process itself), but could be used as basis for continued work on the switch.

Cross Scala support is not natively built into Gradle, but it's possible to use it for same purpose, there are OSS projects like Apache Kafka which do use Gradle to publish artifact variants across supported Scala versions.

I doesn't mean it's not possible, or that nobody is doing it already, just I'm not aware of any project doing cross Scala publishing with Maven. Adding shell script along with the Maven build shows that this use case is not well supported in the tool, fighting with the tool (rigid/opinionated as Maven is) to use it for something it's not designed for IMO is wrong, just use appropriate tool for the job.

**Comment by rawkintrevo** [comment]
I discovered a "maven" based way to do this with profiles.  This isn't to say SBT isn't the long term solution, but I don't know anything about that, and converting everything over is going to be a big task I think (it seems like it was back when that PR/JIRA was opened.)

By adding sub-modules for each Spark, we can still build a given spark/scala combo, but we now also have the `Pall-spark` and soon `-Pall-scala` which will build all relative varients of spark/scala. 

Will finish off the scalas (which will have to be done in `math-scala` and `viennacl`) tomorrow.  I'm not going to add this to Flink unless we add that back in, as it is unnessicary, (we only need the `all-scala` or `all-flink` if we were releasing multiple artifacts.  Since the user must manually build flink anyway, they can simply specify the profile they want.

File: viennacl/scala-2.10/pom.xml [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.mahout</groupId>
+        <artifactId>mahout-native-viennacl</artifactId>
+        <version>0.13.2-SNAPSHOT</version>
```
Comment (andrewpalumbo) [inline viennacl/scala-2.10/pom.xml]: should be 0.13.1-SNAPSHOT

File: viennacl/scala-2.11/pom.xml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.mahout</groupId>
+        <artifactId>mahout-native-viennacl</artifactId>
+        <version>0.13.2-SNAPSHOT</version>
```
Comment (andrewpalumbo) [inline viennacl/scala-2.11/pom.xml]: 0.13.1-SNAPSHOT here as well

File: math-scala/scala-2.11/pom.xml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.mahout</groupId>
+        <artifactId>mahout-math-scala</artifactId>
+        <version>0.13.2-SNAPSHOT</version>
```
Comment (andrewpalumbo) [inline math-scala/scala-2.11/pom.xml]: 0.13.1-SNAPSHOT

**Comment by andrewpalumbo** [comment]
Looks like the version has been updated to  `0.13.2-SNAPSHOT` in a few modules.. see inline comments.. but also in h20.  Will try manually building with this.

**Comment by andrewpalumbo** [comment]
Did a global replace of `0.13.2` for `0.13.1` and same for `0.13.0` -> 0.13.0 so that all modules are versioned as `0.13.1-SNAPSHOT` 
Getting:
```
[INFO] Reactor Summary:
[INFO] 
[INFO] Mahout Build Tools ................................. SUCCESS [  2.357 s]
[INFO] Apache Mahout ...................................... SUCCESS [  0.052 s]
[INFO] Mahout Math ........................................ SUCCESS [ 10.786 s]
[INFO] Mahout HDFS ........................................ SUCCESS [  2.512 s]
[INFO] Mahout Map-Reduce .................................. SUCCESS [ 19.918 s]
[INFO] Mahout Integration ................................. SUCCESS [  4.165 s]
[INFO] Mahout Examples .................................... SUCCESS [ 19.936 s]
[INFO] Mahout Math Scala bindings ......................... SUCCESS [ 44.516 s]
[INFO] Mahout Math Scala 2.10 bindings .................... SUCCESS [  0.409 s]
[INFO] Mahout Release Package ............................. FAILURE [  0.977 s]
[INFO] Mahout Math Scala 2.11 bindings .................... SKIPPED
[INFO] Mahout Spark bindings .............................. SKIPPED
[INFO] Mahout Spark-1.6 Bindings .......................... SKIPPED
[INFO] Mahout Spark-2.0 Bindings .......................... SKIPPED
[INFO] Mahout Spark-2.1 Bindings .......................... SKIPPED
[INFO] Mahout H2O backend ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:46 min
[INFO] Finished at: 2017-07-26T13:48:21-07:00
[INFO] Final Memory: 83M/1107M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-remote-resources-plugin:1.5:process (default) on project apache-mahout-distribution: Failed to resolve dependencies for one or more projects in the reactor. Reason: Missing:
[ERROR] ----------
[ERROR] 1) org.apache.mahout:mahout-spark_2.10:jar:spark_1.6:0.13.1-SNAPSHOT
[ERROR] 
[ERROR] Try downloading the file manually from the project website.
[ERROR] 
[ERROR] Then, install it using the command:
[ERROR] mvn install:install-file -DgroupId=org.apache.mahout -DartifactId=mahout-spark_2.10 -Dversion=0.13.1-SNAPSHOT -Dclassifier=spark_1.6 -Dpackaging=jar -Dfile=/path/to/file
[INFO] Reactor Summary:
[INFO] 
[INFO] Mahout Build Tools ................................. SUCCESS [  2.357 s]
[INFO] Apache Mahout ...................................... SUCCESS [  0.052 s]
[INFO] Mahout Math ........................................ SUCCESS [ 10.786 s]
[INFO] Mahout HDFS ........................................ SUCCESS [  2.512 s]
[INFO] Mahout Map-Reduce .................................. SUCCESS [ 19.918 s]
[INFO] Mahout Integration ................................. SUCCESS [  4.165 s]
[INFO] Mahout Examples .................................... SUCCESS [ 19.936 s]
[INFO] Mahout Math Scala bindings ......................... SUCCESS [ 44.516 s]
[INFO] Mahout Math Scala 2.10 bindings .................... SUCCESS [  0.409 s]
[INFO] Mahout Release Package ............................. FAILURE [  0.977 s]
[INFO] Mahout Math Scala 2.11 bindings .................... SKIPPED
[INFO] Mahout Spark bindings .............................. SKIPPED
[INFO] Mahout Spark-1.6 Bindings .......................... SKIPPED
[INFO] Mahout Spark-2.0 Bindings .......................... SKIPPED
[INFO] Mahout Spark-2.1 Bindings .......................... SKIPPED
[INFO] Mahout H2O backend ................................. SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 01:46 min
[INFO] Finished at: 2017-07-26T13:48:21-07:00
[INFO] Final Memory: 83M/1107M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-remote-resources-plugin:1.5:process (default) on project apache-mahout-distribution: Failed to resolve dependencies for one or more projects in the reactor. Reason: Missing:
[ERROR] ----------
[ERROR] 1) org.apache.mahout:mahout-spark_2.10:jar:spark_1.6:0.13.1-SNAPSHOT
[ERROR] 
[ERROR] Try downloading the file manually from the project website.
[ERROR] 
[ERROR] Then, install it using the command:
[ERROR] mvn install:install-file -DgroupId=org.apache.mahout -DartifactId=mahout-spark_2.10 -Dversion=0.13.1-SNAPSHOT -Dclassifier=spark_1.6 -Dpackaging=jar -Dfile=/path/to/file
```

**Comment by rawkintrevo** [comment]
Took a lot more blind fire on this than I'm proud of. @andrewmusselman @andrewpalumbo and anyone else. Please give this a shot at your convenience.

`mvn clean package -Pall-scala,all-spark,viennacl`

Should yield jars for h2o- scala 2.10/1, viennaccl- scala 2.10/1, spark-1.6-scala-2.10, spark 2.0-scala-2.11,spark-2.1-scala-2.11

**Comment by rawkintrevo** [comment]
@andrewpalumbo @andrewmusselman @pferrel 
Please see my last comment and test this at your convenience. I think we have a winner.

**Comment by andrewpalumbo** [comment]
:+1: will be able to check out likely tomorrow.

**Comment by andrewpalumbo** [comment]
built and tested with `mvn clean package -Pall-scala,all-spark,viennacl`  

I think we may have had issues with deployment.. would like to try running a shell or something but IMO this is close to bring ready to merge to master..

viennacl tests are not kicking being discovered still but @rawkintrevo has fixed this in h2o (was having the same issue before,ie. tests not being discovered).  

I am +1 on merging and then fixing separate issued in smaller PRs before this becomes unweildly.  

Would like to wait on @pferrel 's input though.  (just discussed on slack and he is testing for his current needs) that will probably be a better gauge.

**Comment by andrewpalumbo** [comment]
BTW- we may need to change with the Jenkins config up a bit.

**Comment by pferrel** [comment]
Has anyone tried linking to an application? I get duplicate versions of things with multiple scala versions attached to some things. 

```
[INFO] [Engine$] [error] Modules were resolved with conflicting cross-version suffixes in {file:/Users/pat/ur/}ur:
[INFO] [Engine$] [error]    org.apache.mahout:mahout-math-scala _2.11, _2.10
[INFO] [Engine$] [error]    com.github.scopt:scopt _2.11, _2.10
[INFO] [Engine$] java.lang.RuntimeException: Conflicting cross-version suffixes in: org.apache.mahout:mahout-math-scala, com.github.scopt:scopt
```

When only asking for the 2.11 jars (afaik). I guess I can exclude 2.10 or shade them and may have some other build problem

It would be nice if someone could say they have built an app with mvn or sbt

But +1 for merge with master since it works better than the current master

**Comment by andrewpalumbo** [comment]
@rawkintrevo Looks like we're good to go on this.  Thanks for this.. Please push at your convenience.

**Comment by rawkintrevo** [comment]
per MAHOUT-2020 this method wasn't working with SBT. So I have been refactoring. 

Note, now `mvn clean install` builds all sparks. `mvn clean -Pspark1.6` will result in quicker build times and only build spark 1.6, this is probably required so make the console work.

**Comment by rawkintrevo** [comment]
please see https://github.com/apache/mahout/pull/350

**Pull request closed without merging**